File: SortAndSearch.java
//Сортировка выбором 

public class SelectionSort {
    public static void selectionSort(int[] arr) {
        // Проходим по всем элементам массива, кроме последнего
        for (int i = 0; i < arr.length - 1; i++) {
            // Предполагаем, что минимальный элемент находится на текущей позиции i
            int minIndex = i;
            
            // Ищем минимальный элемент в оставшейся части массива
            for (int j = i + 1; j < arr.length; j++) {
                // Если находим элемент меньше текущего минимального
                if (arr[j] < arr[minIndex]) {
                    // Обновляем индекс минимального элемента
                    minIndex = j;
                }
            }
            
            // Если минимальный элемент не на своем месте, меняем их местами
            if (minIndex != i) {
                int temp = arr[i];      // Сохраняем текущий элемент во временной переменной
                arr[i] = arr[minIndex]; // Заменяем текущий элемент минимальным
                arr[minIndex] = temp;   // Заменяем минимальный элемент сохраненным
            }
        }
    }
}

//сортировка обменом 

public class BubbleSort {
    public static void bubbleSort(int[] arr) {
        int n = arr.length; // Получаем длину массива
        
        // Внешний цикл - количество проходов по массиву
        for (int i = 0; i < n - 1; i++) {
            // Внутренний цикл - сравнение соседних элементов
            for (int j = 0; j < n - i - 1; j++) {
                // Если текущий элемент больше следующего
                if (arr[j] > arr[j + 1]) {
                    // Меняем элементы местами
                    int temp = arr[j];     // Сохраняем текущий элемент
                    arr[j] = arr[j + 1];   // Заменяем текущий элемент следующим
                    arr[j + 1] = temp;     // Заменяем следующий элемент сохраненным
                }
            }
        }
    }
}

//сортировка вставками 

public class InsertionSort {
    public static void insertionSort(int[] arr) {
        int n = arr.length; // Получаем длину массива
        
        // Начинаем со второго элемента (индекс 1)
        for (int i = 1; i < n; i++) {
            int key = arr[i]; // Запоминаем текущий элемент для вставки
            int j = i - 1;    // Начинаем сравнение с предыдущего элемента
            
            // Сдвигаем элементы большие key вправо
            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j]; // Сдвигаем элемент вправо
                j = j - 1;           // Переходим к следующему элементу слева
            }
            
            // Вставляем key на правильную позицию
            arr[j + 1] = key;
        }
    }
}

//сортировка слиянием 

public class MergeSort {
    // Основной метод сортировки
    public static void mergeSort(int[] arr, int left, int right) {
        // Если левая граница меньше правой (есть что сортировать)
        if (left < right) {
            // Находим среднюю точку для разделения массива
            int mid = left + (right - left) / 2;
            
            // Рекурсивно сортируем левую половину
            mergeSort(arr, left, mid);
            
            // Рекурсивно сортируем правую половину
            mergeSort(arr, mid + 1, right);
            
            // Объединяем отсортированные половины
            merge(arr, left, mid, right);
        }
    }
    
    // Метод для слияния двух отсортированных массивов
    private static void merge(int[] arr, int left, int mid, int right) {
        // Вычисляем размеры временных массивов
        int n1 = mid - left + 1; // Размер левого подмассива
        int n2 = right - mid;    // Размер правого подмассива
        
        // Создаем временные массивы
        int[] leftArr = new int[n1];
        int[] rightArr = new int[n2];
        
        // Копируем данные во временные массивы
        for (int i = 0; i < n1; i++) {
            leftArr[i] = arr[left + i];
        }
        for (int j = 0; j < n2; j++) {
            rightArr[j] = arr[mid + 1 + j];
        }
        
        // Индексы для временных массивов и основного массива
        int i = 0, j = 0, k = left;
        
        // Сливаем временные массивы обратно в основной массив
        while (i < n1 && j < n2) {
            // Выбираем меньший элемент из двух массивов
            if (leftArr[i] <= rightArr[j]) {
                arr[k] = leftArr[i]; // Берем элемент из левого массива
                i++;                 // Переходим к следующему элементу левого массива
            } else {
                arr[k] = rightArr[j]; // Берем элемент из правого массива
                j++;                  // Переходим к следующему элементу правого массива
            }
            k++; // Переходим к следующей позиции в основном массиве
        }
        
        // Копируем оставшиеся элементы левого массива (если есть)
        while (i < n1) {
            arr[k] = leftArr[i];
            i++;
            k++;
        }
        
        // Копируем оставшиеся элементы правого массива (если есть)
        while (j < n2) {
            arr[k] = rightArr[j];
            j++;
            k++;
        }
    }
}

//сортировка Шелла

public class ShellSort {
    public static void shellSort(int[] arr) {
        int n = arr.length; // Получаем длину массива
        
        // Начинаем с большого промежутка, затем уменьшаем его
        for (int gap = n / 2; gap > 0; gap /= 2) {
            // Выполняем сортировку вставками для этого промежутка
            for (int i = gap; i < n; i++) {
                // Сохраняем arr[i] во временной переменной
                int temp = arr[i];
                int j;
                
                // Сдвигаем элементы, пока не найдем правильную позицию для arr[i]
                for (j = i; j >= gap && arr[j - gap] > temp; j -= gap) {
                    arr[j] = arr[j - gap]; // Сдвигаем элемент на gap позиций
                }
                
                // Вставляем сохраненный элемент в правильную позицию
                arr[j] = temp;
            }
        }
    }
}

//быстрая сортировка 

public class QuickSort {
    // Основной метод быстрой сортировки
    public static void quickSort(int[] arr, int low, int high) {
        // Если есть что сортировать (low < high)
        if (low < high) {
            // Разделяем массив и получаем индекс опорного элемента
            int pi = partition(arr, low, high);
            
            // Рекурсивно сортируем элементы до опорного
            quickSort(arr, low, pi - 1);
            
            // Рекурсивно сортируем элементы после опорного
            quickSort(arr, pi + 1, high);
        }
    }
    
    // Метод для разделения массива
    private static int partition(int[] arr, int low, int high) {
        // Выбираем последний элемент как опорный
        int pivot = arr[high];
        
        // Индекс меньшего элемента (указывает на правильную позицию опорного)
        int i = low - 1;
        
        // Проходим по всем элементам от low до high-1
        for (int j = low; j < high; j++) {
            // Если текущий элемент меньше или равен опорному
            if (arr[j] <= pivot) {
                i++; // Увеличиваем индекс меньшего элемента
                
                // Меняем местами arr[i] и arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        
        // Помещаем опорный элемент на правильную позицию
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        
        // Возвращаем позицию опорного элемента
        return i + 1;
    }
}

//пирамидальная сортировка 

public class HeapSort {
    public static void heapSort(int[] arr) {
        int n = arr.length; // Получаем длину массива
        
        // Построение max-heap (перегруппировка массива)
        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(arr, n, i);
        }
        
        // Извлекаем элементы из кучи один за другим
        for (int i = n - 1; i > 0; i--) {
            // Перемещаем текущий корень в конец
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;
            
            // Вызываем heapify на уменьшенной куче
            heapify(arr, i, 0);
        }
    }
    
    // Преобразование поддерева в max-heap
    private static void heapify(int[] arr, int n, int i) {
        int largest = i;        // Инициализируем наибольший элемент как корень
        int left = 2 * i + 1;   // Левый дочерний элемент
        int right = 2 * i + 2;  // Правый дочерний элемент
        
        // Если левый дочерний элемент больше корня
        if (left < n && arr[left] > arr[largest]) {
            largest = left;
        }
        
        // Если правый дочерний элемент больше текущего наибольшего
        if (right < n && arr[right] > arr[largest]) {
            largest = right;
        }
        
        // Если наибольший элемент не корень
        if (largest != i) {
            // Меняем местами корень и наибольший элемент
            int swap = arr[i];
            arr[i] = arr[largest];
            arr[largest] = swap;
            
            // Рекурсивно преобразуем затронутое поддерево
            heapify(arr, n, largest);
        }
    }
}

//последовательный поиск 

public class LinearSearch {
    public static int linearSearch(int[] arr, int target) {
        // Проходим по всем элементам массива
        for (int i = 0; i < arr.length; i++) {
            // Если нашли целевой элемент
            if (arr[i] == target) {
                return i; // Возвращаем индекс найденного элемента
            }
        }
        
        // Если элемент не найден, возвращаем -1
        return -1;
    }
}

//бинарный поиск 

public class BinarySearch {
    public static int binarySearch(int[] arr, int target) {
        int left = 0;                    // Левая граница поиска
        int right = arr.length - 1;      // Правая граница поиска
        
        // Пока левая граница не превысила правую
        while (left <= right) {
            int mid = left + (right - left) / 2; // Находим середину
            
            // Если элемент в середине равен целевому
            if (arr[mid] == target) {
                return mid; // Возвращаем индекс
            }
            
            // Если целевой элемент меньше среднего
            if (arr[mid] > target) {
                right = mid - 1; // Ищем в левой половине
            } else {
                left = mid + 1;  // Ищем в правой половине
            }
        }
        
        // Если элемент не найден
        return -1;
    }
}

//интерполирующий поиск 

public class InterpolationSearch {
    public static int interpolationSearch(int[] arr, int target) {
        int low = 0;                    // Нижняя граница
        int high = arr.length - 1;      // Верхняя граница
        
        // Пока целевой элемент находится между границами и границы не совпали
        while (low <= high && target >= arr[low] && target <= arr[high]) {
            // Если границы совпали
            if (low == high) {
                // Если элемент найден, возвращаем индекс
                if (arr[low] == target) {
                    return low;
                }
                return -1; // Иначе возвращаем -1
            }
            
            // Вычисляем позицию с помощью интерполяционной формулы
            int pos = low + ((target - arr[low]) * (high - low)) / (arr[high] - arr[low]);
            
            // Если элемент найден
            if (arr[pos] == target) {
                return pos; // Возвращаем позицию
            }
            
            // Если целевой элемент меньше элемента в позиции pos
            if (arr[pos] > target) {
                high = pos - 1; // Ищем в левой части
            } else {
                low = pos + 1;  // Ищем в правой части
            }
        }
        
        // Если элемент не найден
        return -1;
    }
}

//поиск Фибоначчи 

public class FibonacciSearch {
    public static int fibonacciSearch(int[] arr, int target) {
        int n = arr.length; // Длина массива
        
        // Инициализируем числа Фибоначчи
        int fibMMm2 = 0;          // (m-2)-е число Фибоначчи
        int fibMMm1 = 1;          // (m-1)-е число Фибоначчи
        int fibM = fibMMm2 + fibMMm1; // m-е число Фибоначчи
        
        // Находим наименьшее число Фибоначчи, большее или равное n
        while (fibM < n) {
            fibMMm2 = fibMMm1;
            fibMMm1 = fibM;
            fibM = fibMMm2 + fibMMm1;
        }
        
        int offset = -1; // Отметка об исключенном диапазоне
        
        // Пока есть элементы для проверки
        while (fibM > 1) {
            // Проверяем, valid ли fibMMm2
            int i = Math.min(offset + fibMMm2, n - 1);
            
            // Если target больше значения в i, отсекаем подмассив от offset до i
            if (arr[i] < target) {
                fibM = fibMMm1;
                fibMMm1 = fibMMm2;
                fibMMm2 = fibM - fibMMm1;
                offset = i;
            }
            // Если target меньше значения в i, отсекаем подмассив после i
            else if (arr[i] > target) {
                fibM = fibMMm2;
                fibMMm1 = fibMMm1 - fibMMm2;
                fibMMm2 = fibM - fibMMm1;
            }
            // Элемент найден
            else {
                return i;
            }
        }
        
        // Сравниваем последний элемент
        if (fibMMm1 == 1 && arr[offset + 1] == target) {
            return offset + 1;
        }
        
        // Элемент не найден
        return -1;
    }
}

File: SortAndSearch.py
# Сортировка выбором
def selection_sort(array):
    n = len(array)
    
    for i in range(n - 1):
        min_index = i
        
        for j in range(i + 1, n):
            if array[j] < array[min_index]:
                min_index = j
        
        array[i], array[min_index] = array[min_index], array[i]

def selection_sort_demo():
    example_array = [64, 25, 12, 22, 11]
    
    print("Исходный массив:")
    print_array(example_array)
    
    selection_sort(example_array)
    
    print("\nОтсортированный массив:")
    print_array(example_array)


# Сортировка пузырьком
def bubble_sort(array):
    n = len(array)
    
    for i in range(n - 1):
        swapped = False
        
        for j in range(n - i - 1):
            if array[j] > array[j + 1]:
                array[j], array[j + 1] = array[j + 1], array[j]
                swapped = True
        
        if not swapped:
            break

def bubble_sort_demo():
    example_array = [64, 34, 25, 12, 22, 11]
    
    print("Исходный массив:")
    print_array(example_array)
    
    bubble_sort(example_array)
    
    print("\nОтсортированный массив:")
    print_array(example_array)


# Сортировка вставками
def insertion_sort(array):
    n = len(array)
    
    for i in range(1, n):
        key = array[i]
        j = i - 1
        
        while j >= 0 and array[j] > key:
            array[j + 1] = array[j]
            j -= 1
        
        array[j + 1] = key

def insertion_sort_demo():
    example_array = [12, 11, 13, 5, 6]
    
    print("Исходный массив:")
    print_array(example_array)
    
    insertion_sort(example_array)
    
    print("\nОтсортированный массив:")
    print_array(example_array)


# Сортировка слиянием
def merge_sort(array):
    if len(array) <= 1:
        return array
    
    mid = len(array) // 2
    left_half = array[:mid]
    right_half = array[mid:]
    
    left_half = merge_sort(left_half)
    right_half = merge_sort(right_half)
    
    return merge(left_half, right_half)

def merge(left, right):
    result = []
    i = j = 0
    
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    
    result.extend(left[i:])
    result.extend(right[j:])
    
    return result

def merge_sort_demo():
    example_array = [38, 27, 43, 3, 9, 82, 10]
    
    print("Исходный массив:")
    print_array(example_array)
    
    sorted_array = merge_sort(example_array)
    
    print("\nОтсортированный массив:")
    print_array(sorted_array)


# Сортировка Шелла
def shell_sort(array):
    n = len(array)
    
    gap = n // 2
    while gap > 0:
        for i in range(gap, n):
            temp = array[i]
            j = i
            
            while j >= gap and array[j - gap] > temp:
                array[j] = array[j - gap]
                j -= gap
            
            array[j] = temp
        
        gap //= 2

def shell_sort_demo():
    example_array = [23, 12, 1, 8, 34, 56, 7]
    
    print("Исходный массив:")
    print_array(example_array)
    
    shell_sort(example_array)
    
    print("\nОтсортированный массив:")
    print_array(example_array)


# Быстрая сортировка
def quick_sort(array, low=0, high=None):
    if high is None:
        high = len(array) - 1
    
    if low < high:
        pi = partition(array, low, high)
        
        quick_sort(array, low, pi - 1)
        quick_sort(array, pi + 1, high)

def partition(array, low, high):
    pivot = array[high]
    i = low - 1
    
    for j in range(low, high):
        if array[j] <= pivot:
            i += 1
            array[i], array[j] = array[j], array[i]
    
    array[i + 1], array[high] = array[high], array[i + 1]
    return i + 1

def quick_sort_demo():
    example_array = [10, 7, 8, 9, 1, 5]
    
    print("Исходный массив:")
    print_array(example_array)
    
    quick_sort(example_array)
    
    print("\nОтсортированный массив:")
    print_array(example_array)


# Пирамидальная сортировка
def heap_sort(arr):
    n = len(arr)
    
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)
    
    for i in range(n - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)

def heapify(arr, size, root_index):
    largest = root_index
    left_child = 2 * root_index + 1
    right_child = 2 * root_index + 2
    
    if left_child < size and arr[left_child] > arr[largest]:
        largest = left_child
    
    if right_child < size and arr[right_child] > arr[largest]:
        largest = right_child
    
    if largest != root_index:
        arr[root_index], arr[largest] = arr[largest], arr[root_index]
        heapify(arr, size, largest)

def heap_sort_demo():
    example_array = [12, 11, 13, 5, 6, 7]
    
    print("Исходный массив:")
    print_array(example_array)
    
    heap_sort(example_array)
    
    print("\nОтсортированный массив:")
    print_array(example_array)


# Линейный поиск
def linear_search(array, target):
    for i in range(len(array)):
        if array[i] == target:
            return i
    return -1

def linear_search_demo():
    data = [3, 8, 1, 10, 5]
    key = 10
    result = linear_search(data, key)
    
    if result != -1:
        print(f"Элемент {key} найден на индексе {result}")
    else:
        print(f"Элемент {key} не найден")


# Бинарный поиск
def binary_search(array, target):
    low = 0
    high = len(array) - 1
    
    while low <= high:
        mid = low + (high - low) // 2
        
        if array[mid] == target:
            return mid
        elif array[mid] < target:
            low = mid + 1
        else:
            high = mid - 1
    
    return -1

def binary_search_demo():
    sorted_array = [1, 3, 5, 7, 9, 11]
    search_key = 7
    index = binary_search(sorted_array, search_key)
    
    if index != -1:
        print(f"Значение {search_key} найдено на индексе {index}")
    else:
        print(f"Значение {search_key} не найдено в массиве.")


# Интерполяционный поиск
def interpolation_search(array, target):
    low = 0
    high = len(array) - 1
    
    while low <= high and target >= array[low] and target <= array[high]:
        if low == high:
            if array[low] == target:
                return low
            return -1
        
        pos = low + ((high - low) // (array[high] - array[low])) * (target - array[low])
        pos = min(max(pos, low), high)
        
        if array[pos] == target:
            return pos
        elif array[pos] < target:
            low = pos + 1
        else:
            high = pos - 1
    
    return -1

def interpolation_search_demo():
    sorted_array = [10, 15, 20, 25, 30, 35, 40, 45, 50]
    search_key = 35
    index = interpolation_search(sorted_array, search_key)
    
    if index != -1:
        print(f"Значение {search_key} найдено на индексе {index}")
    else:
        print(f"Значение {search_key} не найдено в массиве.")


# Поиск Фибоначчи
def fibonacci_search(array, x):
    n = len(array)
    
    fib_m2 = 0
    fib_m1 = 1
    fib_m = fib_m2 + fib_m1
    
    while fib_m < n:
        fib_m2 = fib_m1
        fib_m1 = fib_m
        fib_m = fib_m2 + fib_m1
    
    offset = -1
    
    while fib_m > 1:
        i = min(offset + fib_m2, n - 1)
        
        if array[i] < x:
            fib_m = fib_m1
            fib_m1 = fib_m2
            fib_m2 = fib_m - fib_m1
            offset = i
        elif array[i] > x:
            fib_m = fib_m2
            fib_m1 = fib_m1 - fib_m2
            fib_m2 = fib_m - fib_m1
        else:
            return i
    
    if fib_m1 and offset + 1 < n and array[offset + 1] == x:
        return offset + 1
    
    return -1

def fibonacci_search_demo():
    sorted_array = [10, 22, 35, 40, 45, 50, 80, 82, 85, 90, 100]
    x = 85
    result = fibonacci_search(sorted_array, x)
    
    if result != -1:
        print(f"Элемент {x} найден на индексе {result}")
    else:
        print(f"Элемент {x} не найден в массиве.")


# Вспомогательная функция для вывода массива
def print_array(arr):
    print(' '.join(map(str, arr)))


# Демонстрация всех алгоритмов
if __name__ == "__main__":
    print("=== СОРТИРОВКА ВЫБОРОМ ===")
    selection_sort_demo()
    
    print("\n=== СОРТИРОВКА ПУЗЫРЬКОМ ===")
    bubble_sort_demo()
    
    print("\n=== СОРТИРОВКА ВСТАВКАМИ ===")
    insertion_sort_demo()
    
    print("\n=== СОРТИРОВКА СЛИЯНИЕМ ===")
    merge_sort_demo()
    
    print("\n=== СОРТИРОВКА ШЕЛЛА ===")
    shell_sort_demo()
    
    print("\n=== БЫСТРАЯ СОРТИРОВКА ===")
    quick_sort_demo()
    
    print("\n=== ПИРАМИДАЛЬНАЯ СОРТИРОВКА ===")
    heap_sort_demo()
    
    print("\n=== ЛИНЕЙНЫЙ ПОИСК ===")
    linear_search_demo()
    
    print("\n=== БИНАРНЫЙ ПОИСК ===")
    binary_search_demo()
    
    print("\n=== ИНТЕРПОЛЯЦИОННЫЙ ПОИСК ===")
    interpolation_search_demo()
    
    print("\n=== ПОИСК ФИБОНАЧЧИ ===")
    fibonacci_search_demo()